// ep_timestamp/static/js/index.js
var _, $, jQuery;
var $ = require('ep_etherpad-lite/static/js/rjquery').$;
var _ = require('ep_etherpad-lite/static/js/underscore');
//var Security = require('ep_etherpad-lite/static/js/security.js'); 

var padEditor;
var tags = ['timestamp'];
exports.aceRegisterBlockElements = function(){
  return tags;
}

exports.timestamp = {
	initialize: function(context){}
}

exports.aceAttribsToClasses = function(name, context) { 
	if (context.key == 'timestamp' && context.value) {
		return ['timestamp:' + context.value];
	}
}

exports.aceDomLineProcessLineAttributes = function(name, context) {
	var cls = context.cls;
	var domline = context.domline;
	var timestamp = /(?:^| )timestamp:([0-9]*)/.exec(cls);
	var time_cls = ''; 
	
	if (timestamp){
		time_cls = timestamp[0].replace('timestamp:', 'created_');
	}
	var modifier = {
		preHtml: '<div class="line'+time_cls+'">',
		postHtml: '</div>',
		processedMarker: true
	};
	return [modifier];
}

exports.aceInitialized = function(name, context){
  var editorInfo = context.editorInfo;
  padEditor = context.editorInfo.editor;
}
exports.postAceInit = function(name, context){
	exports.timestamp.initialize(context);
}
